### PR TITLE
Local up cluster enhancements

### PR DESCRIPTION
**Pull request by pohly**
**What this PR does / why we need it**:

Networking in a cluster brought up by hack/local-up-cluster.sh was broken because kube-proxy fails to start: connectivity to the Internet is okay, but anything involving communication among services (for example, starting cluster DNS) fails.

Debugging this took a while because there was no indication in the output of hack/local-up-cluster.sh that something had gone wrong. There were also some harmless "no such process" errors during cleanup.

The kube-proxy startup issue was fixed in parallel by PR #60431 but that fix turned out to be incomplete, so this PR uses a different approach, suppresses the harmless error messages (caused by a race) and also makes it more obvious when processes die by checking them regularly.

**Special notes for your reviewer**:

I used

     KUBE_ENABLE_CLUSTER_DNS=true NET_PLUGIN=kubenet hack/local-up-cluster.sh -O

to start the cluster. Check the kube-dns pod and its logs to see how kube-dns fails to communicate with the API server.

I've left out health checking of kubelet when run under Docker because that isn't something that I have used myself yet and I wanted to send the enhancements without further delays. I can try to add that later once you agree that such a health check is useful.

**Release note**:
```release-note
fix network setup in hack/local-up-cluster.sh (https://github.com/kubernetes/kubernetes/pull/60431)
better error diagnostics in hack/local-up-cluster.sh output
```

### CODE REVIEW

**Comment by dims** [comment]
@pohly the kube-proxy problem was fixed already https://github.com/kubernetes/kubernetes/pull/60431 i believe. yes, the other 2 commits in this PR are valuable for sure!

**Comment by pohly** [comment]
On Thu, 2018-03-01 at 01:42 -0800, Davanum Srinivas wrote:
> @pohly the kube-proxy problem was fixed already #60431 i believe. 

Right. I had rebased onto master before submitting, but because that
went cleanly missed that someone had edited the file in the meantime.
What are the chances that this is broken for so long and then two
people debug and fix it concurrently? ;-}

#60431 looks like a cleaner fix, so I'll just remove my change.

> yes, the other 2 commits in this PR are valuable for sure!

Let's keep them then in this PR.

Force-pushed and updated PR description (in particular the release
notes).

**Comment by dims** [comment]
@pohly Please see https://github.com/kubernetes/kubernetes/pull/60431#issuecomment-369624579 apparently i was wrong in my patch. So please resurrect yours :) (Thanks @ncdc)

**Comment by pohly** [comment]
On Thu, 2018-03-01 at 15:54 +0000, Davanum Srinivas wrote:
> @pohly Please see #60431 (comment) apparently i was wrong in my
> patch. So please resurrect yours :) (Thanks @ncdc)

"git reflog" to the rescue... ;-)

I've re-added a slightly updated commit. Please have another look.

kube-proxy seems to pick up the feature gates correctly:

$ FEATURE_GATES="MountPropagation=true,CSIPersistentVolume=true" hack/local-up-cluster.sh -O
...

$ grep CSIPersistent /tmp/kube-proxy.log 
I0301 17:10:49.026534   44807 feature_gate.go:226] feature gates: &{map[CSIPersistentVolume:true MountPropagation:true]}
$ cat /tmp/kube-proxy.yaml 
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
clientConnection:
  kubeconfig: /var/run/kubernetes/kube-proxy.kubeconfig
hostnameOverride: 127.0.0.1
mode: 
featureGates:
  MountPropagation: true
  CSIPersistentVolume: true

**Comment by dims** [comment]
LGTM 👍

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -400,6 +400,44 @@ cleanup()
   exit 0
 }
 
+# Check if all processes are still running. Prints a warning once each time
+# a process dies unexpectedly.
+healthcheck()
```
Comment (ixdy) [inline hack/local-up-cluster.sh]: minor nit: for consistency with the rest of the file, use `function healthcheck {` instead.

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -400,6 +400,44 @@ cleanup()
...
+{
+  if [[ -n "${APISERVER_PID-}" ]] && ! sudo kill -0 ${APISERVER_PID} 2>/dev/null; then
+    warning "API server terminated unexpectedly, see ${APISERVER_LOG}"
+    APISERVER_PID=
+  fi
+
+  if [[ -n "${CTRLMGR_PID-}" ]] && ! sudo kill -0 ${CTRLMGR_PID} 2>/dev/null; then
```
Comment (ixdy) [inline hack/local-up-cluster.sh]: this seems to actually be `$CTLRMGR_PID` (and `$CTLRMGR_LOG` below)

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -400,6 +400,44 @@ cleanup()
...
+    warning "kube-controller-manager terminated unexpectedly, see ${CTRLMGR_LOG}"
+    CTRLMGR_PID=
+  fi
+
+  if [[ -n "$DOCKERIZE_KUBELET" ]]; then
+    # TODO (?): check health also in this case.
```
Comment (ixdy) [inline hack/local-up-cluster.sh]: maybe create an issue and link it here?

File: hack/local-up-cluster.sh [hunk 4]
```diff
@@ -400,6 +400,44 @@ cleanup()
...
+    :
```
Comment (ixdy) [inline hack/local-up-cluster.sh]: nit - stray `:`?

File: hack/local-up-cluster.sh [hunk 5]
```diff
@@ -400,6 +400,44 @@ cleanup()
...
+    KUBELET_PID=
+  fi
+
+  if [[ -n "${PROXY_PID-}" ]] && ! sudo kill -0 ${PROXY_PID} 2>/dev/null; then
+    warning "kube-proxy terminated unexpectedly, see ${PROXY_LOG}"
+    PROXY_PID=
+  fi
+
+  if [[ -n "${SCHEDULER_PID-}" ]] && ! sudo kill -0 ${SCHEDULER_PID} 2>/dev/null; then
+    warning "API server terminated unexpectedly, see ${SCHEDULER_LOG}"
```
Comment (ixdy) [inline hack/local-up-cluster.sh]: scheduler, not API server?

**Review by ixdy — COMMENTED** [review]
a few minor nits, otherwise LG

File: hack/local-up-cluster.sh (hunk 4 above)
Comment (pohly) [reply hack/local-up-cluster.sh]: No, the : is intentional. A statement is needed between "then" and "fi", and : is the shortest possible statement that doesn't do anything.

File: hack/local-up-cluster.sh (hunk 2 above)
Comment (pohly) [reply hack/local-up-cluster.sh]: Right, good catch. Fixed.

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (pohly) [reply hack/local-up-cluster.sh]: Fixed.

File: hack/local-up-cluster.sh (hunk 3 above)
Comment (pohly) [reply hack/local-up-cluster.sh]: Done, see https://github.com/kubernetes/kubernetes/issues/62474

File: hack/local-up-cluster.sh (hunk 5 above)
Comment (pohly) [reply hack/local-up-cluster.sh]: Darn copy-and-paste ;-} You are right of course.

**Comment by pohly** [comment]
@ixdy thanks for your review. I think I have addressed your concerns in the rebased and force-pushed commits.

**Comment by pohly** [comment]
@dims and/or @ixdy: after some retries we currently have a clean test status. Can you help with getting this PR merged?

**Pull request merged**